src/smartpeak/source/ui/Heatmap2DWidget.cpp

```diff
@@ -6,14 +6,169 @@ namespace SmartPeak
+    std::vector<std::string> feature_names;
+    for (int i = 0; i < selected_feature_names.size(); ++i) {
+      if (std::count(feature_names.begin(), feature_names.end(), selected_feature_names(i)) == 0 && !selected_feature_names(i).empty())
+        feature_names.push_back(selected_feature_names(i));
+    }
+    if (std::find(feature_names.begin(), feature_names.end(), selected_feature_) == feature_names.end())
+    {
+      if (feature_names.size())
+      {
+        selected_feature_ = feature_names.front();
+      }
+      else
+      {
+        selected_feature_ = "";
+      }
+    }
     // Main graphic
-    if (rows_.size() > 1 || columns_.size() > 1) {
-      assert(data_.dimension(0) == rows_.size() && data_.dimension(1) == columns_.size());
-      const ImPlotFlags imPlotFlags = ImPlotFlags_MousePos | ImPlotFlags_Highlight | ImPlotFlags_BoxSelect | ImPlotFlags_ContextMenu;
-      if (ImPlot::BeginPlot(plot_title_.c_str(), x_axis_title_.c_str(), y_axis_title_.c_str(), ImVec2(plot_width_ - 25, plot_height_ - 40), imPlotFlags)) {
-        ImPlot::PlotHeatmap(("##" + plot_title_).c_str(), data_.data(), rows_.size(), columns_.size(), data_min_, data_max_, NULL);
-        ImPlot::EndPlot();
+    if (selected_feature_.empty())
+    {
+      ImGui::Text("No features to select");
+    }
+    else
+    {
+      if (ImGui::BeginCombo("Select feature", selected_feature_.c_str()))
+      {
+        for (const auto& feature_name : feature_names)
+        {
+          if (ImGui::Selectable(feature_name.c_str()))
+          {
+            selected_feature_ = feature_name;
+          }
+        }
+        ImGui::EndCombo();
+      }
+
+      // Check if we need to refresh the data
+      Eigen::Tensor<std::string, 1> selected_sample_names = session_handler_.getSelectSampleNamesPlot();
+      Eigen::Tensor<std::string, 1> selected_transitions = session_handler_.getSelectTransitionsPlot();
+      Eigen::Tensor<std::string, 1> selected_transition_groups = session_handler_.getSelectTransitionGroupsPlot();
+      if ((refresh_needed_) || // data changed
+          (selected_feature_ != heatmap_data_.selected_feature_) ||
+          !compareInput(selected_sample_names, heatmap_data_.selected_sample_names_) ||
+          !compareInput(selected_transitions, heatmap_data_.selected_transitions_) ||
+          !compareInput(selected_transition_groups, heatmap_data_.selected_transition_groups_))
+      {
+        session_handler_.getHeatMap(sequence_handler_, heatmap_data_, selected_feature_);
+        // We need to handle "invalid" data - infinite values, or vey high - which will crash ImPlot.
+        invalid_data_ = false;
+        for (int i = 0; i < heatmap_data_.feat_heatmap_row_labels.size(); ++i)
+        {
+          for (int j = 0; j < heatmap_data_.feat_heatmap_col_labels.size(); ++j)
+          {
+            int index = i * heatmap_data_.feat_heatmap_row_labels.size() + j;
+            if (heatmap_data_.feat_heatmap_data.data()[index] > 1e20)
```

Comment:
If 1e20 is the limit you found for the high value data, I would recommend making this a global constraint that can be shared among all of the plotting widgets.  Perhaps as a base class member?

Reply:
yes, it's a value id decided on my side really, by experiment. I will move it to some constant.